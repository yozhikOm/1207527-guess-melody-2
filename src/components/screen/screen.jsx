import React from 'react';
import PropTypes from "prop-types";
import {WelcomeScreen} from '../welcome-screen/welcome-screen.jsx';
import {GenreQuestionScreen} from '../genre-question-screen/genre-question-screen.jsx';
import {ArtistQuestionScreen} from '../artist-question-screen/artist-question-screen.jsx';
import {GameOverScreen} from '../game-over-screen/game-over-screen.jsx';
import withActivePlayer from '../../hocs/with-active-player/with-active-player.js';
import withUserAnswer from '../../hocs/with-user-answer/with-user-answer.js';

const GenreQuestionScreenWrapped = withActivePlayer(withUserAnswer(GenreQuestionScreen));
const ArtistQuestionScreenWrapped = withActivePlayer(ArtistQuestionScreen);

const Screen = ({gameSettings, question, mistakes, step, onWelcomeScreenClick, onUserAnswer}) => {
  const {
    gameTime,
    errorCount,
  } = gameSettings;

  if (step === -2) {
    return <GameOverScreen
      onStartButtonClick={onWelcomeScreenClick}
    />;
  }

  // if (!question) {
  if (step === -1) {
    return <WelcomeScreen
      gameTime={gameTime}
      errorCount={errorCount}
      onStartButtonClick={onWelcomeScreenClick}
    />;
  }

  switch (question.type) {
    case `genre`: return <GenreQuestionScreenWrapped

      question={question}
      onAnswer={(userAnswer) => onUserAnswer(
          userAnswer,
          question,
          mistakes,
          errorCount
      )}
    />;

    case `artist`: return <ArtistQuestionScreenWrapped
      question={question}
      onAnswer={(userAnswer) => onUserAnswer(
          userAnswer,
          question,
          mistakes,
          errorCount
      )}
    />;
  }

  return null;
};

Screen.propTypes = {
  gameSettings: PropTypes.shape({
    gameTime: PropTypes.number,
    errorCount: PropTypes.number,
  }),
  question: PropTypes.shape({
    type: PropTypes.string.isRequired,
    genre: PropTypes.string,
    answers: PropTypes.arrayOf(
        PropTypes.shape({
          src: PropTypes.string,
          genre: PropTypes.string,
        }).isRequired
    ).isRequired,
  }),
  onWelcomeScreenClick: PropTypes.func.isRequired,
  onUserAnswer: PropTypes.func,
  mistakes: PropTypes.number.isRequired,
  step: PropTypes.number.isRequired,
};

export {Screen};
