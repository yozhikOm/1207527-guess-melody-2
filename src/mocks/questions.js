export const questions = [
  {
    type: `genre`,
    genre: `rock`,
    answers: [
      {
        src: `http://www.hochmuth.com/mp3/Vivaldi_Sonata_eminor_.mp3`,
        genre: `classic`,
      },
      {
        src: `https://files.freemusicarchive.org/storage-freemusicarchive-org/music/no_curator/Scott_Holmes/Inspiring__Upbeat_Music/Scott_Holmes_-_05_-_Little_Idea.mp3`,
        genre: `pop`,
      },
      {
        src: `https://files.freemusicarchive.org/storage-freemusicarchive-org/music/no_curator/KieLoKaz/Free_Ganymed/KieLoKaz_-_03_-_Wow_Kielokaz_ID_359.mp3`,
        genre: `jazz`,
      },
      {
        src: `https://files.freemusicarchive.org/storage-freemusicarchive-org/music/ccCommunity/XTaKeRuX/The_way_iam/XTaKeRuX_-_04_-_Lost.mp3`,
        genre: `rock`,
      },
    ],
  },
  {
    type: `genre`,
    genre: `folk`,
    answers: [
      {
        src: `https://files.freemusicarchive.org/storage-freemusicarchive-org/music/Ziklibrenbib/November_Polaroid/st/November_Polaroid_-_05_-_And_Miles_to_Go.mp3`,
        genre: `folk`,
      },
      {
        src: `https://files.freemusicarchive.org/storage-freemusicarchive-org/music/ccCommunity/Chad_Crouch/Motion/Chad_Crouch_-_03_-_Creative_Writing.mp3`,
        genre: `pop`,
      },
      {
        src: `https://files.freemusicarchive.org/storage-freemusicarchive-org/music/no_curator/Checkie_Brown/hey/Checkie_Brown_-_12_-_Rosalie_CB_29.mp3`,
        genre: `jazz`,
      },
      {
        src: `https://files.freemusicarchive.org/storage-freemusicarchive-org/music/no_curator/Lobo_Loco/Salad_Mixed/Lobo_Loco_-_08_-_Sierra_Tracks_ID_1232.mp3`,
        genre: `folk`,
      },
    ],
  },
  {
    type: `artist`,
    song: {
      artist: `Checkie Brown`,
      src: `https://files.freemusicarchive.org/storage-freemusicarchive-org/music/no_curator/Checkie_Brown/hey/Checkie_Brown_-_08_-_Hippie_Bulle_-Stoned_Funghi_CB_28.mp3`,
    },
    answers: [
      {
        picture: `http://placehold.it/134x134`,
        artist: `John Snow`,
      },
      {
        picture: `http://placehold.it/134x134`,
        artist: `Jack Daniels`,
      },
      {
        picture: `http://placehold.it/134x134`,
        artist: `Checkie Brown`,
      },
    ],
  },
];

