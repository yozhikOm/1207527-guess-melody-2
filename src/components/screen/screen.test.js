import React from 'react';
import {Screen} from './screen.jsx';
import renderer from 'react-test-renderer';
import {questions} from '../../mocks/questions.js';

it(`Screen компонент рендерится корректно`, () => {
  const mockSettings = {
    gameTime: 5,
    errorCount: 3
  };
  const mockQuestion1 = questions.find((q) => q.type === `genre`);
  const mockQuestion2 = questions.find((q) => q.type === `artist`);

  const mockButtonClick = jest.fn();

  const mockMistakes = 0;

  const createNodeMock = () => {
    return {
      oncanplaythrough: null,
      onplay: null,
      onpause: null,
      ontimeupdate: null,
      src: ``
    };
  };

  const screenComponent1 = renderer.create(
      <Screen
        gameSettings={mockSettings}
        question={mockQuestion1}
        mistakes={mockMistakes}
        step={1}
        onUserAnswer={mockButtonClick}
        onWelcomeScreenClick={mockButtonClick}
      />,
      {createNodeMock}
  ).toJSON();
  expect(screenComponent1).toMatchSnapshot();

  const screenComponent2 = renderer.create(
      <Screen
        gameSettings={mockSettings}
        question={mockQuestion2}
        mistakes={mockMistakes}
        step={2}
        onUserAnswer={mockButtonClick}
        onWelcomeScreenClick={mockButtonClick}
      />,
      {createNodeMock}
  ).toJSON();
  expect(screenComponent2).toMatchSnapshot();

  const screenComponent3 = renderer.create(
      <Screen
        gameSettings={mockSettings}
        question={undefined}
        mistakes={mockMistakes}
        step={-1}
        onUserAnswer={mockButtonClick}
        onWelcomeScreenClick={mockButtonClick}
      />,
      {createNodeMock}
  ).toJSON();
  expect(screenComponent3).toMatchSnapshot();
});
