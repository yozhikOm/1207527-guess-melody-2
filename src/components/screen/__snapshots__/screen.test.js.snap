// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen компонент рендерится корректно 1`] = `
<section
  className="game__screen"
>
  <h2
    className="game__title"
  >
    Выберите 
    rock
     треки
  </h2>
  <form
    className="game__tracks"
    onSubmit={[Function]}
  >
    <div
      className="track"
    >
      <button
        className="track__button track__button--play"
        disabled={true}
        onClick={[Function]}
        type="button"
      />
      <div
        className="track__status"
      >
        <audio />
      </div>
      <div
        className="game__answer"
      >
        <input
          className="game__input visually-hidden"
          id="answer-0"
          name="answer"
          onChange={[Function]}
          type="checkbox"
          value="answer-0"
        />
        <label
          className="game__check"
          htmlFor="answer-0"
        >
          Отметить
        </label>
      </div>
    </div>
    <div
      className="track"
    >
      <button
        className="track__button track__button--play"
        disabled={true}
        onClick={[Function]}
        type="button"
      />
      <div
        className="track__status"
      >
        <audio />
      </div>
      <div
        className="game__answer"
      >
        <input
          className="game__input visually-hidden"
          id="answer-1"
          name="answer"
          onChange={[Function]}
          type="checkbox"
          value="answer-1"
        />
        <label
          className="game__check"
          htmlFor="answer-1"
        >
          Отметить
        </label>
      </div>
    </div>
    <div
      className="track"
    >
      <button
        className="track__button track__button--play"
        disabled={true}
        onClick={[Function]}
        type="button"
      />
      <div
        className="track__status"
      >
        <audio />
      </div>
      <div
        className="game__answer"
      >
        <input
          className="game__input visually-hidden"
          id="answer-2"
          name="answer"
          onChange={[Function]}
          type="checkbox"
          value="answer-2"
        />
        <label
          className="game__check"
          htmlFor="answer-2"
        >
          Отметить
        </label>
      </div>
    </div>
    <div
      className="track"
    >
      <button
        className="track__button track__button--play"
        disabled={true}
        onClick={[Function]}
        type="button"
      />
      <div
        className="track__status"
      >
        <audio />
      </div>
      <div
        className="game__answer"
      >
        <input
          className="game__input visually-hidden"
          id="answer-3"
          name="answer"
          onChange={[Function]}
          type="checkbox"
          value="answer-3"
        />
        <label
          className="game__check"
          htmlFor="answer-3"
        >
          Отметить
        </label>
      </div>
    </div>
    <button
      className="game__submit button"
      type="submit"
    >
      Ответить
    </button>
  </form>
</section>
`;

exports[`Screen компонент рендерится корректно 2`] = `
<section
  className="game__screen"
>
  <h2
    className="game__title"
  >
    Кто исполняет эту песню?
  </h2>
  <div
    className="game__track"
  >
    <div
      className="track"
    >
      <button
        className="track__button track__button--play"
        disabled={true}
        onClick={[Function]}
        type="button"
      />
      <div
        className="track__status"
      >
        <audio />
      </div>
    </div>
  </div>
  <form
    className="game__artist"
    onChange={[Function]}
  >
    <div
      className="artist"
    >
      <input
        className="artist__input visually-hidden"
        id="answer-0"
        name="answer"
        type="radio"
        value="answer-0"
      />
      <label
        className="artist__name"
        htmlFor="answer-0"
      >
        <img
          alt="John Snow"
          className="artist__picture"
          src="http://placehold.it/134x134"
        />
        John Snow
      </label>
    </div>
    <div
      className="artist"
    >
      <input
        className="artist__input visually-hidden"
        id="answer-1"
        name="answer"
        type="radio"
        value="answer-1"
      />
      <label
        className="artist__name"
        htmlFor="answer-1"
      >
        <img
          alt="Jack Daniels"
          className="artist__picture"
          src="http://placehold.it/134x134"
        />
        Jack Daniels
      </label>
    </div>
    <div
      className="artist"
    >
      <input
        className="artist__input visually-hidden"
        id="answer-2"
        name="answer"
        type="radio"
        value="answer-2"
      />
      <label
        className="artist__name"
        htmlFor="answer-2"
      >
        <img
          alt="Checkie Brown"
          className="artist__picture"
          src="http://placehold.it/134x134"
        />
        Checkie Brown
      </label>
    </div>
  </form>
</section>
`;

exports[`Screen компонент рендерится корректно 3`] = `
<section
  className="welcome"
>
  <div
    className="welcome__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <button
    className="welcome__button"
    onClick={[MockFunction]}
  >
    <span
      className="visually-hidden"
    >
      Начать игру
    </span>
  </button>
  <h2
    className="welcome__rules-title"
  >
    Правила игры
  </h2>
  <p
    className="welcome__text"
  >
    Правила просты:
  </p>
  <ul
    className="welcome__rules-list"
  >
    <li>
      За 
      00 минут 05 секунд
       нужно ответить на все вопросы.
    </li>
    <li>
      Можно допустить 
      3
       ошибки.
    </li>
  </ul>
  <p
    className="welcome__text"
  >
    Удачи!
  </p>
</section>
`;
